livemerge: prune bitmaps before measuring

aesteve-rh opened the pull request:
Prune stale bitmaps before a live merge to avoid
failing with ENOSPC.

Fixes: #352
Signed-off-by: Albert Esteve <aesteve@redhat.com>

aesteve-rh commented:
/ost

aesteve-rh commented:
Currently I am trying to verify the fix with a stress test. But I post the PR already to refine it early, so that we can integrate as soon as I manage to reproduce the issue (if so).

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2857,6 +2858,25 @@ def measure(self, sdUUID, imgUUID, volUUID, dest_format, backing=True,
+    @public
+    def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
+        """
+        Prune stale bitmaps from the base volume.
```
Please document that this is an internal that should be used only from virt code.

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2857,6 +2858,25 @@ def measure(self, sdUUID, imgUUID, volUUID, dest_format, backing=True,
+                pruned if missing or invalid in top.
+        """
+        base_vol = self._produce_volume(sdUUID, imgUUID, baseUUID)
+        top_vol = self._produce_volume(sdUUID, imgUUID, volUUID)
```
Using _produce_volume() is nice, but it only a shared lock on the storage domain.

We should to take also:
- shared lock on the image? (check how storage jobs lock similar operations)
- shared lock on the top volume
- exclusive lock on the base volume (modified using this code)

Since this should be called only from virt code, and the vm owns the volumes,
no other storage code should access them (engine should ensure this) but it
is safer to take locks so bad engine request will fail at vdsm instead of
corrupting the volume.

You can add locking in a different commit to simplify the work.

nirs commented on lib/vdsm/virt/livemerge.py:
```diff
@@ -287,6 +287,13 @@ def merge(self, driveSpec, base, top, bandwidth, job_id):
             except JobExistsError as e:
                 raise exception.MergeFailed(str(e), job=job.id)
 
+            # Prune stale bitmaps from base volume before measuring.
```
This should be called once when starting a merge - good.

But cannot be used when:
- top is the leaf volume - active layer merge, since top can have bitmaps which did not appear in the image yet. New bitamps created when the vm is running will be written to the image only when the vm stop or when we create a snapshot.
- base is not using qcow2 format, prune will fail

So I think we need something like:

    if self._base_needs_prune_bitmaps(drive, job, base_info):
        self._dom._vm.prune_bitmaps(...)

nirs commented on lib/vdsm/virt/vm.py:
```diff
@@ -6106,6 +6106,16 @@ def onDeviceRemoved(self, device_alias):
+        """
+        Prune stale bitmaps from the base volume.
+        """
+        self.cif.irs.prune_bitmaps(
```
You are not checking errors - this will not raise on errors!

Check how error are handled in measure() bellow.

aesteve-rh replied:
It will, if it fails it will raise an `exception.RemoveBitmapError` from the bitmaps module [1].

[1] https://github.com/oVirt/vdsm/blob/master/lib/vdsm/storage/bitmaps.py#L122

nirs replied:
The exception is not raised in the caller. HSM is wrapped by Dispatcher, which warps public methods to return status dict instead of raising.

aesteve-rh replied:
Ah! I see it now, I though `res` in measure was the return of the hsm measure (dict from the parsed json), but now I see the dispatcher in the decorator. Got it.

nirs commented on tests/virt/vmfakelib.py:
```diff
@@ -51,6 +51,9 @@ def appropriateDevice(self, guid, vmid):
         pass
 
+    def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
+        pass
```
This should use `@mayfail` so we can simulate errors and test merge behavior on failures.

It should also record the fact that this was called and the called arguments, we have
a `@recorded` decorator for this, see how it is used. This will allow verifying that we
prone bitmaps when starting merge for internal merge, and not pruning bitmaps for
active layer merge or internal merge to raw base.

Check how livemerge_test.py verify the flows. We need to add some more asserts in the
existing tests.

aesteve-rh replied:
I saw how `@maybefail` works, is a handy wrapper!

But in this case I think is better fit to use the `response.error_raw` if we want to simulate errors, something like what we do in `getVolumeSize` below, to have similar code to this one:
```python
    @recorded
    def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
        if "prune_bitmaps" in self.errors:
            error = self.errors["prune_bitmaps"]
            return response.error_raw(error.code, error.msg)
        return response.success(result=None)
```

nirs replied:
VM.prune_bitmaps() should not return status code! It should raise on errors. Its purpose is to covert the (horrible) self.irs.cif.prune_bitmaps() call returning status codes to proper call raising on errors.

aesteve-rh replied:
It does raise on errors. This `prune_bitmaps` is in the fake IRS class, not in the VM.

So it returns a status code that is processed by `VM.prune_bitmaps()`, and then raise the error.

The fact that both have the same name may cause some confusion.

nirs replied:
The fake IRS should be wrapped with a Dispatcher so it behaves like the real cif.irs object.

But if the current code is not using a dispatcher and try to behave like a wrapped hsm
and return status codes, you should do the same.

nirs reviewed (changes requested):
To simplify this work, you can split the commits:
- Add internal api to hsm
- Use the new api in live merge
- Update the tests to verify that we prune bitmaps only when we should

Current pr should be good enough to test the happy case for internal merge - you should be able to reproduce this by adding bitmaps manually to running vm base volume.

aesteve-rh commented:
/ost

aesteve-rh commented:
/ost

aesteve-rh commented:
/ost

aesteve-rh commented:
I tried to reproduce the issue, both manually and with ovirt-stress, but I couldn't get to see the "No space left on device" in the libvirtd logs on a full setup with two hosts and a running VM. Unfortunately, I can not spend more time with it.

So I would finish the review of this PR, and integrate it. Hopefully on qe they have better luck reproducing.

nirs commented:
> I tried to reproduce the issue, both manually and with ovirt-stress, but I couldn't get to see the "No space left on device" in the libvirtd logs on a full setup with two hosts and a running VM.

How did you try to reproduce?

When we start live merge we measure the top volume to get the required size of base. This
measure the data and bitmaps in base.
https://github.com/oVirt/vdsm/blob/6f67a73dac7c02009c0125a59e002590f9cc2beb/lib/vdsm/virt/livemerge.py#L405

If base size is smaller than the measured data, we don't extend base.
https://github.com/oVirt/vdsm/blob/6f67a73dac7c02009c0125a59e002590f9cc2beb/lib/vdsm/virt/livemerge.py#L457

So we need to find how much data we must have in top so the measured size is exactly
base size. Then add eough stale bitmaps to base (we needed 6 bitmaps in the tests to
reproduce this).

You can do the measurements with the reproducer script, using 2.5g images (instead of 128m).
When you find the right size, do:

1. Add second disk to running vm using qcow2 format
   - This disk will start as 2.5g empty disk
2. Create a snapshot (snap1)
   - This will create empty 2.5g snapshot
3. Fill this snapshot with the right amount of data so merge should succeed
   - Can be done inside the vm by writing random data directly to the second disk
     ```
     dd if=/dev/urandom bs=1M count=xxx of=/dev/vdb oflag=direct conv=fsync
     ```
     writing zeros will be faster but when committing the data qemu detects
     zeros and convert them to zero clusters, which will prevent reproduction
     of this issue.
4. The snapshot will be extend to 5g during the write - this is fine and does
   not affect the test
5. Create another snapshot (snap2)
   - now we can merge the middle layer into the base
6. Add enough stale bitmaps using `qemu-img bitmap add` to the base volume
7. In engine UI, delete snapshot snap1
   - this merges contents of the middle layer (snap2) into the base layer (snap1)
   - it should fail because of the stale bitmaps

I think we need to reproduce this first before we modify the code. If you don't have
time, maybe someone from QE has time to reproduce.

aesteve-rh commented:
It is exactly how I tried.

First I tried with 3GiB disk, but was not fully provisioned. Vdsm's created image was 2.5GiB, but virtual size was 3GiB, so I changed it to 2.5GiB (2560MiB). Didn't get it either, and I even tried with 128MiB as is what we used for cold merges. No luck.

Tried to write directly in the VM (to be fair I did only try with `/dev/zero`, so today I retried with `/dev/urandom`), or with `qemu-io` in the host directly in the top volume:
```shell
qemu-io -f qcow2 {top_path} -c "write 0 2000M"
qemu-io -f qcow2 {top_path} -c "write 2000M 558M"
```

I did try also with more snapshots, removing the first or the second snapshot (these tests helped understand the process better, actually), automatically and manually.

None of these really worked. At least in what my goal was: I didn't get the qemu error in `journactl` output in any case.

I will document the steps and hand over to QE, I really can not put much more time into trying. I wanted to move on completely today, but still dedicated the morning to try with the `/dev/urandom` proposal, with no luck either.

My feeling with the patch is that it should work, same as we fixed the cold merges: basically runs the same operation as with cold merges, and the test proves that it runs when required. If OST runs pass (it does not degrade), I'm good with it.

nirs commented:
> It is exactly how I tried.
> 
> First I tried with 3GiB disk, but was not fully provisioned. Vdsm's created image was 2.5GiB, but virtual size was 3GiB, so I changed it to 2.5GiB (2560MiB). Didn't get it either, and I even tried with 128MiB as is what we used for cold merges. No luck.
> 
> Tried to write directly in the VM (to be fair I did only try with `/dev/zero`, 

This cannot work, qemu detects the zeros during the merge and does not write any
data, so merge cannot fail.

... 
> None of these really worked. At least in what my goal was: I didn't get the qemu error in `journactl` output in any case.

Getting the error is not reliable way to detect the issue. The issue is detected by
checking the merge logs. When the merge complete after an error, the volume is not
removed from the chain, and engine will report the the snapshot is illegal.

When this happens, trying to delete the snapshot again is expected to fail forever.

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2871,6 +2871,9 @@ def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
         """
+        vars.task.getSharedLock(STORAGE, imgUUID)
+        vars.task.getSharedLock(STORAGE, volUUID)
+        vars.task.getExclusiveLock(STORAGE, baseUUID)
```
This not how storage locks are taken, see how we lock the image and the volume in the sp.py.

nirs replied:
So more specific advice - we cannot take volume locks, since locking the image take a lock on all volumes:

I think it works like this:
1. sd.py _registerResourceNamespaces register imageResourcesNamespace with imageResourceFactory
2. imageResourceFactory takes a lock on all volume of the image when creating a resource
3. So when you take a lock on an image, you lock all the volume in this image with the same lock type

Since we want an exclusive lock on the base volume, we must take exclusive lock on the image, which will take also exclusive lock on the top volume.

So we need something like:

    img_ns = rm.getNamespace(sc.IMAGE_NAMESPACE, sdUUID)
    with rm.acquireResource(img_ns, imgUUID, rm.EXCLUSIVE):
        purne bitmaps here...

BUT: this must be AFTER we take the storage domain lock, otherwise we can have a deadlock. All locks must be taken in the same order:

1. sc.STORAGE
2. sc.IMAGE_NAMESPACE
3. sc.VOLUME_NAMESPACE
4. sc.LVM_ACTIVATION_NAMESPACE

This is why the constants are named: "00_storage", "01_img", "02_vol", "03_lvm".

Code taking multiple locks must sort the locks to ensure the order.

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2871,6 +2871,9 @@ def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
+        vars.task.getSharedLock(STORAGE, volUUID)
+        vars.task.getExclusiveLock(STORAGE, baseUUID)
         base_vol = self._produce_volume(sdUUID, imgUUID, baseUUID)
         top_vol = self._produce_volume(sdUUID, imgUUID, volUUID)
```
This does not work, since we _produce_volume takes a shared lock on the volume, and this code tries to take the same lock twice:

    def _produce_volume(self, sdUUID, imgUUID, volUUID):
        vars.task.getSharedLock(STORAGE, sdUUID)
        ...

So code is very clean but incorrect. What we want to do is to have list of locks, and lock of of them at the same time and in the right order using guarded.context().

    img_ns = rm.getNamespace(sc.IMAGE_NAMESPACE, self.sd_id)
    locks = [rm.Lock(sc.STORAGE, sdUUID, rm.SHARED),
             rm.Lock(img_ns, imgUUID, rm.EXCLUSIVE)]
    with guarded.context(locks):
        prune bitmaps here...

You need to produce the volume manually instead of using the _produce_volume() helper, it does too much and cannot be used when using multiple volumes. Maybe it can be fixed to support this usage but lets keep this change simple and avoid unneeded changes that can break unrelated flows and need much more testing.

nirs reviewed (changes requested):
Partial review about locking. I will review the rest later this week.

aesteve-rh commented:
I have finally been able to ***consistently*** reproduce the issue with 2560MiB disks.

# Reproducing steps

1. Create snap1
2. Write 2555MiB from the running VM:
```
dd if=/dev/urandom bs=1M count=2555 of=/dev/vda oflag=direct conv=fsync
```
3. Create snap2
4. Add 70 bitmaps to base image from the Host:
```
for i in $(seq 70); do
    qemu-img bitmap --add /dev/<vg_name>/<lv_name> stale-bitmap-$i
done
```
Results in:
Top image measure:
```
required size: 2679767040
fully allocated size: 2685009920
bitmaps size: 0
```
Base image measure:
```
required size: 655360
fully allocated size: 2685009920
bitmaps size: 9240576
```
5. Remove snap1 from engine UI to trigger the internal merge.

# Error case (without the fix in this PR)

Vdsm logs:
```
2022-12-05 18:21:39,245+0100 ERROR (libvirt/events) [virt.vm] (vmId='bb1ff201-a159-4436-a23f-8dad4cced1e9') Block job 89667ef9-0b0f-4b63-8ed9-9a720e0d4845 type COMMIT for drive vdb has failed (vm:5955)
```
Resulting base measure:
```
required size: 2675048448
fully allocated size: 2685009920
bitmaps size: 9240576
```
Engine UI:
![image](https://user-images.githubusercontent.com/102953499/206135142-2168814e-552a-4c51-ad3f-b0cdaeae3389.png)

If we skip adding the bitmaps the merge succeeds, so it is an error caused by the stale bitmaps.

# Success case (with the fix in this PR):
Vdsm logs.

After `START merge` and before measuring the relevant disk, we see the prune_bitmaps logs:
```
2022-12-07 09:49:47,864+0100 WARN  (jsonrpc/2) [storage.bitmaps] Prune stale bitmaps ['stale-bitmap-1', 'stale-bitmap-2', 'stale-bitmap-3', 'stale-bitmap-4', 'stale-bitmap-5', 'stale-bitmap-6', 'stale-bitmap-7', 'stale-bitmap-8', 'stale-bitmap-9', 'stale-bitmap-10', 'stale-bitmap-11', 'stale-bitmap-12', 'stale-bitmap-13', 'stale-bitmap-14', 'stale-bitmap-15', 'stale-bitmap-16', 'stale-bitmap-17', 'stale-bitmap-18', 'stale-bitmap-19', 'stale-bitmap-20', 'stale-bitmap-21', 'stale-bitmap-22', 'stale-bitmap-23', 'stale-bitmap-24', 'stale-bitmap-25', 'stale-bitmap-26', 'stale-bitmap-27', 'stale-bitmap-28', 'stale-bitmap-29', 'stale-bitmap-30', 'stale-bitmap-31', 'stale-bitmap-32', 'stale-bitmap-33', 'stale-bitmap-34', 'stale-bitmap-35', 'stale-bitmap-36', 'stale-bitmap-37', 'stale-bitmap-38', 'stale-bitmap-39', 'stale-bitmap-40', 'stale-bitmap-41', 'stale-bitmap-42', 'stale-bitmap-43', 'stale-bitmap-44', 'stale-bitmap-45', 'stale-bitmap-46', 'stale-bitmap-47', 'stale-bitmap-48', 'stale-bitmap-49', 'stale-bitmap-50', 'stale-bitmap-51', 'stale-bitmap-52', 'stale-bitmap-53', 'stale-bitmap-54', 'stale-bitmap-55', 'stale-bitmap-56', 'stale-bitmap-57', 'stale-bitmap-58', 'stale-bitmap-59', 'stale-bitmap-60', 'stale-bitmap-61', 'stale-bitmap-62', 'stale-bitmap-63', 'stale-bitmap-64', 'stale-bitmap-65', 'stale-bitmap-66', 'stale-bitmap-67', 'stale-bitmap-68', 'stale-bitmap-69', 'stale-bitmap-70'] from '/rhev/data-center/mnt/blockSD/a2ad7a23-96c5-4c23-9007-c0d1892ea07d/images/d4e763f5-67d8-4a73-9c3b-79da330032e5/0033c0b7-69f3-43fa-b36c-8c0a518aea84' (bitmaps:144)
2022-12-07 09:49:47,878+0100 INFO  (jsonrpc/2) [storage.bitmaps] Remove bitmap stale-bitmap-1 from '/rhev/data-center/mnt/blockSD/a2ad7a23-96c5-4c23-9007-c0d1892ea07d/images/d4e763f5-67d8-4a73-9c3b-79da330032e5/0033c0b7-69f3-43fa-b36c-8c0a518aea84' (bitmaps:117)
...
```
Vdsm measures expected data:
```
FINISH measure return={'result': {'bitmaps': 0, 'required': 2679767040, 'fully-allocated': 2685009920}} from=::ffff:192.168.122.96,38934, flow_id=13305ae3-0084-4eed-9f5a-ec31a43424ff, task_id=b77b670e-79ae-4cad-8d64-71b6259b83b8 (api:37)
```
And measure finishes correctly, no ERRORs in the logs.

Other disks also have a start `prune_bitmaps`, without the bitmaps removal WARN logs:
```
INFO  (jsonrpc/3) [vdsm.api] START prune_bitmaps(sdUUID='a2ad7a23-96c5-4c23-9007-c0d1892ea07d', imgUUID='c7b5c2f6-87ca-4ba6-a498-e12025f0ab12', volUUID='285874a4-0fc6-4bcf-9afc-696f40dd12bc', baseUUID='7f2f1e6e-d32d-4883-941e-836b77c4dde3') from=::ffff:192.168.122.96,38934, flow_id=13305ae3-0084-4eed-9f5a-ec31a43424ff, task_id=55136c39-aacb-4c02-9d8f-17d93a763b5f (api:31)
...
INFO  (jsonrpc/3) [vdsm.api] FINISH prune_bitmaps return=None from=::ffff:192.168.122.96,38934, flow_id=13305ae3-0084-4eed-9f5a-ec31a43424ff, task_id=55136c39-aacb-4c02-9d8f-17d93a763b5f (api:37)
```

Resulting base measure (bitmaps removed):
```
required size: 2679767040
fully allocated size: 2685009920
bitmaps size: 0
```

Engine logs:
```
INFO  [org.ovirt.engine.core.dal.dbbroker.auditloghandling.AuditLogDirector] (EE-ManagedScheduledExecutorService-engineScheduledThreadPool-Thread-32) [13305ae3-0084-4eed-9f5a-ec31a43424ff] EVENT_ID: USER_REMOVE_SNAPSHOT_FINISHED_SUCCESS(356), Snapshot 'snap 1' deletion for VM 'internal-livemerge-vm' has been completed.
```

Also disappeared in the engine UI, and the task finished successfully.

aesteve-rh commented:
/ost

aesteve-rh commented:
/ost

michalskrivanek commented:
/ost

aesteve-rh commented:
/ost

aesteve-rh commented:
/ost

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -34,6 +34,7 @@
 from vdsm.common.time import monotonic_time
```
For the commit message:
> - exclusive lock on the base image

Remove `base`, image is the disk including base and top.

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2857,6 +2859,33 @@ def measure(self, sdUUID, imgUUID, volUUID, dest_format, backing=True,
+            dom = sdCache.produce_manifest(sdUUID=sdUUID)
+            try:
+                base_vol = dom.produceVolume(imgUUID=imgUUID, volUUID=baseUUID)
+                top_vol = dom.produceVolume(imgUUID=imgUUID, volUUID=volUUID)
```
In produce_manifest() and produceVolume() there no need to use key=value in
such call, both arguments are required, and the key=value style adds no value
because the name of the argument is self documentating.

nirs commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2857,6 +2859,33 @@ def measure(self, sdUUID, imgUUID, volUUID, dest_format, backing=True,
+                base_vol = dom.produceVolume(imgUUID=imgUUID, volUUID=baseUUID)
+                top_vol = dom.produceVolume(imgUUID=imgUUID, volUUID=volUUID)
+            except se.VolumeDoesNotExist as e:
+                raise exception.expected(e)
```
This is good for public API, meaning that this is not vdsm error but an error of the
caller. But in this internal API it is better to raise to make debugging easier, so
we can remove the try-except block.

nirs commented on lib/vdsm/virt/livemerge.py:
```diff
@@ -287,6 +287,14 @@ def merge(self, driveSpec, base, top, bandwidth, job_id):
                 raise exception.MergeFailed(str(e), job=job.id)
 
+            if self._base_needs_prune_bitmaps(drive, job, base_info):
+                # Prune stale bitmaps from base volume before measuring.
```
The comment does not add any value, it describe what the code does, which is pretty 
clear.

The only interesting info is "before measuring". This comment can come before the if
and say something like:

    # Prune stale bitmaps if needed once before measuring the image. Current
    # qemu-img ignores bitmaps in base but future version may not.
    if self._base_needs_prune_bitmaps(drive, job, base_info):
        ...

nirs commented on lib/vdsm/virt/livemerge.py:
```diff
@@ -402,6 +410,21 @@ def _start_commit(self, drive, job):
+        if base_info['format'] != 'COW':
+            log.debug("Base volume does not support bitmap operations "
+                      "job=%s drive=%s volume=%s",
+                      job.id, job.drive, job.base)
```
Base volume does not support bitmaps. I'm not sure that this log is needed.

nirs commented on lib/vdsm/virt/livemerge.py:
```diff
@@ -402,6 +410,21 @@ def _start_commit(self, drive, job):
+                      job.id, job.drive, job.base)
+            return False
+
+        if job.top == drive.volumeID:
```
This should use:

    if job.active_commit:

to match other checks like _base_needs_extend().

aesteve-rh replied:
Ah, I see, it stores the same condition in the `Job` class. Good. I'll use it.

nirs commented on lib/vdsm/virt/livemerge.py:
```diff
@@ -402,6 +410,21 @@ def _start_commit(self, drive, job):
+
+        if job.top == drive.volumeID:
+            log.debug("Top volume is the active volume "
+                      "job=%s drive=%s volume=%s",
```
It would be nice to unify the way we log job and base. You copied the style use in _base_needs_extend() but in other code we log "job_id=%r", "base=%r" or "(job: %s)".

I think "job=%r base=%r" here will be better and later someone will unify the logs.

aesteve-rh replied:
How about this for both debug logs (regarding text & style)?

```python
    def _base_needs_prune_bitmaps(self, job, base_info):
        if base_info['format'] != 'COW':
            log.debug("Base volume does not support bitmaps, "
                      "job=%r base=%r", job.id, job.base)
            return False

        if job.active_commit:
            log.debug("Top volume is the active volume, "
                      "job=%r top=%s", job.id, job.top)
            return False
```

nirs commented on tests/virt/vmfakelib.py:
```diff
@@ -53,6 +54,9 @@ def inappropriateDevices(self, ident):
     def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
+        if "prune_bitmaps" in self.errors:
+            error = self.errors["prune_bitmaps"]
+            return response.error_raw(error.code, error.msg)
```
We can simplify this to:

    error = self.errors.get("prune_bitmaps")
    if error:
        raise ...

This is shorter and does only one dict lookup.

nirs reviewed (commented):
Looks good, see comments for minor logs, comments, and style improvements.

aesteve-rh commented:
/ost

aesteve-rh commented on lib/vdsm/storage/hsm.py:
```diff
@@ -2857,6 +2859,30 @@ def measure(self, sdUUID, imgUUID, volUUID, dest_format, backing=True,
+        locks = [rm.Lock(sc.STORAGE, sdUUID, rm.SHARED),
+                 rm.Lock(img_ns, imgUUID, rm.EXCLUSIVE)]
+        with guarded.context(locks):
+            dom = sdCache.produce_manifest(sdUUID=sdUUID)
```
oops :)
